### ISSUE TITLE
org.postgresql.util.PSQLException: Unexpected SQL query error: cannot cast type time with time zone to timestamp without time zone

### ISSUE THREAD

**Issue by swiss-knight** [interactive-instruments/ldproxy#610]
Hi,

I've got a PostgreSQL table with some "timestamp without time zone" fields. Everything goes well.

And another one with "timestamp with time zone" and also "time with time zone" fields, and here, a `500 Internal Server Error` is raised.

Here's the full stacktrace:

```
DEBUG [2022-02-17 23:57:14,100] service                 - Processing request: GET /collections/collectionIds/items [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,107] service                 - accept */* [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,108] service                 - content-type Optional[ApiMediaType{type=application/geo+json, label=GeoJSON, parameter=json, fileExtension=json, qs=1000}] [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,108] service                 - accept-language null [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,108] service                 - content-language Optional[en] [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,154] service                 - accept */* [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,154] service                 - content-type Optional[ApiMediaType{type=application/problem+json, label=PROBLEM+JSON, parameter=json, fileExtension=json, qs=1000}] [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
ERROR [2022-02-17 23:57:14,155] service                 - Server Error with ID 51dd287bab4517db: IO operation was stopped unexpectedly after 0 bytes because of org.postgresql.util.PSQLException: Unexpected SQL query error: 
  ERROR: cannot cast type time with time zone to timestamp without time zone
  Position: 166 [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
DEBUG [2022-02-17 23:57:14,155] service                 - Stacktrace: [cd1f78b1-b8eb-448e-830c-4bf34f65f438] 
! org.postgresql.util.PSQLException: Unexpected SQL query error: 
!   ERROR: cannot cast type time with time zone to timestamp without time zone
!   Position: 166
! at de.ii.xtraplatform.feature.provider.sql.domain.SqlConnector$1.apply(SqlConnector.java:211)
! at de.ii.xtraplatform.feature.provider.sql.domain.SqlConnector$1.apply(SqlConnector.java:174)
! at akka.japi.JavaPartialFunction.apply(JavaAPI.scala:151)
! at akka.stream.impl.fusing.MapError$$anon$8.onUpstreamFailure(Ops.scala:333)
! at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:525)
! at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:390)
! at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:650)
! at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:521)
! at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:625)
! at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:800)
! at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$shortCircuitBatch(ActorGraphInterpreter.scala:787)
! at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:819)
! at akka.actor.Actor.aroundReceive(Actor.scala:537)
! at akka.actor.Actor.aroundReceive$(Actor.scala:535)
! at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:716)
! at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
! at akka.actor.ActorCell.invoke(ActorCell.scala:548)
! at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
! at akka.dispatch.Mailbox.run(Mailbox.scala:231)
! at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
! at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
! at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
! at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
! at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
! at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
! Causing: akka.stream.IOOperationIncompleteException: IO operation was stopped unexpectedly after 0 bytes because of org.postgresql.util.PSQLException: Unexpected SQL query error: 
!   ERROR: cannot cast type time with time zone to timestamp without time zone
!   Position: 166
! at akka.stream.impl.io.OutputStreamGraphStage$$anon$1.onUpstreamFailure(OutputStreamGraphStage.scala:66)
! at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:525)
! at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:390)
! at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:650)
! at akka.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute(ActorGraphInterpreter.scala:61)
! at akka.stream.impl.fusing.ActorGraphInterpreter$SimpleBoundaryEvent.execute$(ActorGraphInterpreter.scala:57)
! at akka.stream.impl.fusing.ActorGraphInterpreter$BatchingActorInputBoundary$OnError.execute(ActorGraphInterpreter.scala:82)
! at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:625)
! at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:800)
! at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:818)
! at akka.actor.Actor.aroundReceive(Actor.scala:537)
! at akka.actor.Actor.aroundReceive$(Actor.scala:535)
! at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:716)
! at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
! at akka.actor.ActorCell.invoke(ActorCell.scala:548)
! at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
! at akka.dispatch.Mailbox.run(Mailbox.scala:231)
! at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
! at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
! at java.base/java.lang.Thread.run(Unknown Source)
DEBUG [2022-02-17 23:57:14,155] service                 - Sending response: 500 Internal Server Error [cd1f78b1-b8eb-448e-830c-4bf34f65f438]
```

PG database version:    
`"PostgreSQL 13.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 10.2.1_pre1) 10.2.1 20201203, 64-bit"`

ldproxy was run using it's latest docker image.

Doc related to PG datetime data type: https://www.postgresql.org/docs/13/datatype-datetime.html

I'm wondering if I can do anything to fix this error?

Also, feel free to tell me how I can print the underlying SQL query to see what exactly happened. 

Thanks a lot!

**Comment by azahnen** [comment]
Hi, well a time cannot be used as datetime. If the configuration was auto-generated, this might be a bug.

To fix it, just change the type for properties using time columns in the provider yml to `STRING`.

To enable sql query logging, set this in cfg.yml:

```
logging:
  sqlQueries: ON
```

**Comment by swiss-knight** [comment]
Hello @azahnen,

Thanks for the hints!

Yes, indeed, the configuration was auto-generated (I don't exactly remember when, but it was, for sure).

With the SQL queries enabled in the log, I was able to figure out what exactly went wrong.

Here is the relevant part of the enhanced stacktrace:

```
DEBUG [2022-02-18 17:35:00,647] service                 - Executing statement: WITH
    NR AS (SELECT MIN(SKEY) AS minKey, MAX(SKEY) AS maxKey, count(*) AS numberReturned FROM (SELECT A.id AS SKEY FROM localisations A ORDER BY SKEY LIMIT 10) AS IDS),
    NM AS (SELECT count(*) AS numberMatched FROM (SELECT A.id AS SKEY FROM localisations A ORDER BY 1) AS IDS) 
  SELECT * FROM NR, NM [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,692]                         - Execution of prepared statement took 24ms  
DEBUG [2022-02-18 17:35:00,694] service                 - Executing statement: SELECT A.id AS SKEY, A.feature_id, A.user_id, A.admin_id, A.remark, A.date_verified::timestamp(0), A.date_last_seen::timestamp(0), A.date_ref::timestamp(0), A.start::timestamp(0), A.stop::timestamp(0), A.id, ST_AsText(ST_ForcePolygonCCW(A.location_point)), A.date_checked::timestamp(0), ST_AsText(ST_ForcePolygonCCW(A.groundgeom_poly)), A.state FROM localisations A WHERE (A.id >= 1 AND A.id <= 10) ORDER BY 1 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,704] service                 - numberMatched 171501 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,704] service                 - numberReturned 10 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,704] service                 - pageSize 10 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,704] service                 - page 1 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,704] service                 - pages 17151 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,726] service                 - accept text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8 [f884793a-abea-44a7-af8f-488517c4d5c6] 
DEBUG [2022-02-18 17:35:00,726] service                 - content-type Optional[ApiMediaType{type=text/html, label=HTML, parameter=html, fileExtension=html, qs=1000}] [f884793a-abea-44a7-af8f-488517c4d5c6] 
ERROR [2022-02-18 17:35:00,727] service                 - Server Error with ID 94fda77fdd4fe3c6: IO operation was stopped unexpectedly after 0 bytes because of org.postgresql.util.PSQLException: Unexpected SQL query error: 
  ERROR: cannot cast type time with time zone to timestamp without time zone
  Position: 166 [f884793a-abea-44a7-af8f-488517c4d5c6]
```

Data types of all these date/time fields are as follows in the PostgreSQL database:

```
A.date_verified:   timestamp with time zone
A.date_last_seen:  date
A.date_ref:        timestamp with time zone
A.start:           time with time zone
A.stop:            time with time zone
A.date_checked:    timestamp with time zone
```

And indeed, it seems that it tries to cast all these fields to `::timestamp(0)` regardless of their actual data type.


Finally, here are the date fields in the provider yaml file for that `locations` table:
```
      date_verified:
        sourcePath: date_verified
        type: DATETIME
        role: PRIMARY_INSTANT
      date_last_seen:
        sourcePath: date_last_seen
        type: DATETIME
      date_ref:
        sourcePath: date_ref
        type: DATETIME
      start:
        sourcePath: start
        type: DATETIME
      stop:
        sourcePath: stop
        type: DATETIME
      date_checked:
        sourcePath: date_checked
        type: DATETIME
      stop:
        sourcePath: stop
        type: DATETIME
```

I tried to change the `type` of `start`and `stop` to `TIME` and the one of `date_last_seen` to `DATE` but then things went bad during the startup of the container:

```
ERROR [2022-02-18 17:53:18,382]                          - Cannot load 'entities/providers/service.yml': Cannot deserialize value of type `de.ii.xtraplatform.features.domain.SchemaBase$Type` from String "TIME": not one of the values accepted for Enum class: [STRING, GEOMETRY, BOOLEAN, OBJECT_ARRAY, UNKNOWN, DATETIME, FLOAT, DATE, OBJECT, VALUE_ARRAY, INTEGER]
 at [Source: (byte[])"---
id: service
createdAt: 1643145562759
lastModified: 1643145562759
entityStorageVersion: 2
providerType: FEATURE
featureProviderType: SQL
nativeCrs:
  code: 4326
  forceAxisOrder: LON_LAT
connectionInfo:
  database: service
  host: localhost:5432
  user: service
  password: ************************************
types:
  attachments:
    sourcePath: /attachments
    type: OBJECT
    properties:
      feature_id:
        sourcePath: image_i"[truncated 15553 bytes]; line: 172, column: 15] (through reference chain: de.ii.xtraplatform.features.domain.ImmutableFeatureProviderCommonData$Builder["types"]->de.ii.xtraplatform.features.domain.ApplyKeyToValueMap["localisations"]->de.ii.xtraplatform.features.domain.ImmutableFeatureSchema$Builder["properties"]->java.util.LinkedHashMap["start"]->de.ii.xtraplatform.features.domain.ImmutableFeatureSchema$Builder["type"]) 
```
Form the enum here above,  we can see that `TIME` is not recognized as a valid value.

Finally, by replacing the type `TIME` by `STRING`, I successfully managed to make it work.    
But as time is a valid data type in PostgreSQL it's maybe worth it to take it into consideration?

I leave this issue open and let you decide whether to close it once you have found a solution on your side or not.

Thanks!

**Comment by azahnen** [comment]
Will be addressed as part of #346.
